keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2144 @@
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes adding a declarative API to manage a node maintenance. This API can be used to
+implement additional capabilities around node draining.
+
+## Motivation
+The goal of this KEP is to analyze and improve node maintenance in Kubernetes.
+
+Node maintenance is a request from a cluster administrator to remove all pods
+from a node(s) so that it can be disconnected from the cluster to perform a
+software upgrade (OS, Kubelet), hardware upgrade, or simply to remove the node
+as it is no longer needed.
+
+Kubernetes has existing support for this use case in the following way with `kubectl drain`:
+1. There are running pods on node A, some of which are protected with PodDisruptionBudgets (PDB).
+2. Set the node `Unschedulable` (cordon) to prevent new pods from being scheduled there.
+3. Evict (default behavior) pods from node A by using the eviction API (see [kubectl drain worklflow](https://raw.githubusercontent.com/kubernetes/website/f2ef324ac22e5d9378f2824af463777182817ca6/static/images/docs/kubectl_drain.svg)).
+4. Proceed with the maintenance and shut down the node.
+5. Kubelet can try to delay the shutdown to allow the remaining pods to terminate gracefully
```

Comment:
```suggestion
5. Kubelet will try to delay the shutdown to allow the remaining pods to terminate gracefully
```

Reply:
Is that true @kwilczynski ? What's your thinking around using “will” here?

Reply:
@sftim, capability vs possibility. Can it delay the shutdown? Yes. Will it do it? Yes. Provided that the feature is enabled (should be). But, I might be wrong about this. Thoughts?

Reply:
We can't assume that the kubelet is able to delay a shutdown (might be a node which doesn't offer that option).

We could write
```suggestion
5. On nodes that support it, the kubelet may detect the imminent shutdown and then trigger graceful
   shutdown handling:
   * delay the shutdown pending graceful termination of remaining pods
   * terminate remaining pods in reverse priority order
   
   Only some nodes support this graceful shutdown handling.
```

Reply:
We could.

To add more colour:

However, the current implementation of Graceful Node Shutdown relies on underlying node support, assuming that the feature is enabled and the platform supports it. We'd fail if we could not inhibit the shutdown. There will be no graceful shutdown of any sort if we cannot set an inhibitor for whatever reason.

As such, what about this the following:

> On platforms and nodes that support it, the kubelet will try to detect the imminent shutdown and then
> attempt to perform a graceful node shutdown.

Reply:
Thanks for the suggestions, updated